src/librustc/dep_graph/graph.rs

```diff
@@ -710,14 +710,15 @@ impl DepGraph {
                                 return None
                             }
                             None => {
-                                if !tcx.sess.has_errors() {
+                                if !tcx.sess.has_errors_or_delayed_span_bugs() {
                                     bug!("try_mark_previous_green() - Forcing the DepNode \
                                           should have set its color")
                                 } else {
-                                    // If the query we just forced has resulted
-                                    // in some kind of compilation error, we
-                                    // don't expect that the corresponding
```

Comment:
TBH I do not grasp the reasons of this expectation. 
But whatever they are, seems the test is a counter-example?

Reply:
The comment should be re-worded to "If the query we just forced has resulted in some kind of compilation error, we cannot rely on the dep-node color having been properly updated."

Reply:
To elaborate: Forcing a query (as we have done when reaching this point in the code) must update the `colors` table with an entry for that query invocation. Reaching the `None` branch here means that we did not find the expected entry in the table, so we abort compilation with the `bug!`. However, there seem to be cases when something generated an error while forcing the query where table is not updated. This is supposedly harmless (although it would be great to find out how that happens exactly) because once an error has occurred we don't persist anything to the incremental cache.

So the comment should expanded to something like:

```
// If the query we just forced has resulted in 
// some kind of compilation error, we cannot rely on
// the dep-node color having been properly updated.
// This means that the query system has reached an
// invalid state. We let the compiler continue (by
// returning `None`) so it can emit error messages
// and wind down, but rely on the fact that this
// invalid state will not be persisted to the 
// incremental compilation cache because of 
// compilation errors being present.
```

Reply:
Also, the new behavior of returning instead of continuing marking is definitely an improvement!

Reply:
Thanks, that makes sense now!